### PR TITLE
Remove null-check from codegen

### LINKED ISSUE #821: example https://github.com/pulumi/examples/tree/master/aws-java-eks-minimal fails to run
### What happened?

The aws-java-eks-minimal example stopped working.  Downgrading to ```com.pulumi:pulumi 0.5.2``` makes it work again so there is a regression introduced in the Java SDK somewhere between 0.5.2 and 0.5.3. The preview of `aws:ec2/getVpc:getVpc result` hits a null check and fails to deserialize.

### Steps to reproduce

Clone `pulumi/exampes`, `cd aws-java-eks-minimal` and `pulumi preview`

### Expected Behavior

It is supposed to create k8s cluster in aws

### Actual Behavior

```
C:\temp\examples\aws-java-eks-minimal>pu
Previewing update (dev)

View Live: https://app.pulumi.com/nagkumar/eks-minimal/dev/previews/9ce040d0-6d92-41dc-bdfd-d83caa400ff9

     Type                 Name             Plan       Info
 +   pulumi:pulumi:Stack  eks-minimal-dev  create     2 errors; 44 messages

Diagnostics:
  pulumi:pulumi:Stack (eks-minimal-dev):
    Aug 29, 2022 3:39:19 PM com.pulumi.deployment.internal.DeploymentImpl$DefaultEngineLogger errorAsync
    SEVERE: Running program [PID: 2796](unknown) failed with an unhandled exception:
    java.lang.UnsupportedOperationException: Convert [aws:ec2/getVpc:getVpc result]: Error converting 'com.google.common.collect.RegularImmutableMap' to 'TypeShape{type=class com.pulumi.aws.ec2.outputs.GetVpcResult, parameters=[]}'. Unexpected exception: null
        at com.pulumi.serialization.internal.Converter.convertObjectUntyped(Converter.java:119)
        at com.pulumi.serialization.internal.Converter.convertValue(Converter.java:86)
        at com.pulumi.deployment.internal.DeploymentImpl$Invoke.parseInvokeResponse(DeploymentImpl.java:482)
        at com.pulumi.deployment.internal.DeploymentImpl$Invoke.lambda$invoke$1(DeploymentImpl.java:473)
        at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
        at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073)
        at net.javacrumbs.futureconverter.guavacommon.GuavaFutureUtils$ListenableFutureBackedValueSourceFuture$1.onSuccess(GuavaFutureUtils.java:139)
        at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1138)
        at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:31)
        at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1270)
        at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:1038)
        at com.google.common.util.concurrent.AbstractFuture.set(AbstractFuture.java:783)
        at io.grpc.stub.ClientCalls$GrpcFuture.set(ClientCalls.java:558)
        at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:531)
        at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:553)
        at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:68)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:739)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:718)
        at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
    Caused by: java.lang.IllegalStateException: Unexpected exception: null
        at com.pulumi.serialization.internal.Converter.lambda$tryConvertObjectInner$7(Converter.java:450)
        at java.base/java.util.HashMap.forEach(HashMap.java:1336)
        at com.pulumi.serialization.internal.Converter.tryConvertObjectInner(Converter.java:409)
        at com.pulumi.serialization.internal.Converter.convertObjectUntyped(Converter.java:115)
        ... 23 more
    Caused by: java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at com.pulumi.serialization.internal.Converter.lambda$tryConvertObjectInner$7(Converter.java:441)
        ... 26 more
    Caused by: java.lang.NullPointerException
        at java.base/java.util.Objects.requireNonNull(Objects.java:221)
        at com.pulumi.aws.ec2.outputs.GetVpcResult$Builder.state(GetVpcResult.java:291)
        ... 31 more

    error: Running program [PID: 2796](unknown) failed with an unhandled exception:
    java.lang.UnsupportedOperationException: Convert [aws:ec2/getVpc:getVpc result]: Error converting 'com.google.common.collect.RegularImmutableMap' to 'TypeShape{type=class com.pulumi.aws.ec2.outputs.GetVpcResult, parameters=[]}'. Unexpected exception: null
        at com.pulumi.serialization.internal.Converter.convertObjectUntyped(Converter.java:119)
        at com.pulumi.serialization.internal.Converter.convertValue(Converter.java:86)
        at com.pulumi.deployment.internal.DeploymentImpl$Invoke.parseInvokeResponse(DeploymentImpl.java:482)
        at com.pulumi.deployment.internal.DeploymentImpl$Invoke.lambda$invoke$1(DeploymentImpl.java:473)
        at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
        at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073)
        at net.javacrumbs.futureconverter.guavacommon.GuavaFutureUtils$ListenableFutureBackedValueSourceFuture$1.onSuccess(GuavaFutureUtils.java:139)
        at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1138)
        at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:31)
        at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1270)
        at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:1038)
        at com.google.common.util.concurrent.AbstractFuture.set(AbstractFuture.java:783)
        at io.grpc.stub.ClientCalls$GrpcFuture.set(ClientCalls.java:558)
        at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:531)
        at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:553)
        at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:68)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:739)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:718)
        at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
    Caused by: java.lang.IllegalStateException: Unexpected exception: null
        at com.pulumi.serialization.internal.Converter.lambda$tryConvertObjectInner$7(Converter.java:450)
        at java.base/java.util.HashMap.forEach(HashMap.java:1336)
        at com.pulumi.serialization.internal.Converter.tryConvertObjectInner(Converter.java:409)
        at com.pulumi.serialization.internal.Converter.convertObjectUntyped(Converter.java:115)
        ... 23 more
    Caused by: java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at com.pulumi.serialization.internal.Converter.lambda$tryConvertObjectInner$7(Converter.java:441)
        ... 26 more
    Caused by: java.lang.NullPointerException
        at java.base/java.util.Objects.requireNonNull(Objects.java:221)
        at com.pulumi.aws.ec2.outputs.GetVpcResult$Builder.state(GetVpcResult.java:291)
        ... 31 more
    error: an unhandled error occurred: 'd:\Apps\Dev\langs\java\bldeng\mvn\bin\mvn.cmd d:\Apps\Dev\langs\java\bldeng\mvn\bin\mvn.cmd -Dorg.slf4j.simpleLogger.defaultLogLevel=warn --no-transfer-progress compile exec:java' exited with non-zero exit code: 1
```

### Output of `pulumi about`

```
CLI          
Version      3.38.0
Go Version   go1.19
Go Compiler  gc

Plugins
NAME        VERSION
aws         5.13.0
eks         0.37.1
java        unknown
kubernetes  3.22.0-alpha.1661453084+94c160dd

Host     
OS       darwin
Version  12.5.1
Arch     x86_64

This project is written in java: executable='/Users/anton/.nix-profile/bin/java' version='openjdk 11.0.11 2021-04-20 LTS
OpenJDK Runtime Environment Zulu11.48+21-CA (build 11.0.11+9-LTS)
OpenJDK 64-Bit Server VM Zulu11.48+21-CA (build 11.0.11+9-LTS, mixed mode)' maven='Apache Maven 3.8.4 (9b656c72d54e5bacbed989b64718c159fe39b537)' gradle='7.4' java='/Users/anton/.nix-profile/bin/java' javac='11.0.11'

Current Stack: dev1

Found no resources associated with dev1

Found no pending operations associated with dev1

Backend        
Name           pulumi.com
URL            https://app.pulumi.com/t0yv0
User           t0yv0
Organizations  t0yv0, pulumi

No dependencies found

Pulumi locates its logs in /var/folders/mb/s1jx27m10t3082fqclpgytt80000gn/T/ by default
```

### Additional context

```mvn dependency:tree```

```
[INFO] Scanning for projects...
[INFO] 
[INFO] ------------------< com.pulumi:aws-java-eks-minimal >-------------------
[INFO] Building aws-java-eks-minimal 1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-dependency-plugin:2.8:tree (default-cli) @ aws-java-eks-minimal ---
[WARNING] The artifact xml-apis:xml-apis:jar:2.0.2 has been relocated to xml-apis:xml-apis:jar:1.0.b2
[INFO] com.pulumi:aws-java-eks-minimal:jar:1.0-SNAPSHOT
[INFO] +- com.pulumi:pulumi:jar:0.5.4:compile
[INFO] |  +- io.grpc:grpc-protobuf:jar:1.38.0:runtime
[INFO] |  |  +- io.grpc:grpc-api:jar:1.38.0:runtime
[INFO] |  |  |  \- io.grpc:grpc-context:jar:1.38.0:runtime
[INFO] |  |  +- com.google.api.grpc:proto-google-common-protos:jar:2.0.1:runtime
[INFO] |  |  +- io.grpc:grpc-protobuf-lite:jar:1.38.0:runtime
[INFO] |  |  +- com.google.errorprone:error_prone_annotations:jar:2.4.0:compile
[INFO] |  |  \- org.codehaus.mojo:animal-sniffer-annotations:jar:1.19:runtime
[INFO] |  +- io.grpc:grpc-stub:jar:1.38.0:runtime
[INFO] |  +- com.google.code.findbugs:jsr305:jar:3.0.2:compile
[INFO] |  +- net.javacrumbs.future-converter:future-converter-java8-guava:jar:1.2.0:runtime
[INFO] |  |  +- net.javacrumbs.future-converter:future-converter-common:jar:1.2.0:runtime
[INFO] |  |  +- net.javacrumbs.future-converter:future-converter-java8-common:jar:1.2.0:runtime
[INFO] |  |  \- net.javacrumbs.future-converter:future-converter-guava-common:jar:1.2.0:runtime
[INFO] |  +- com.google.guava:guava:jar:31.1-jre:compile
[INFO] |  |  +- com.google.guava:failureaccess:jar:1.0.1:compile
[INFO] |  |  +- com.google.guava:listenablefuture:jar:9999.0-empty-to-avoid-conflict-with-guava:compile
[INFO] |  |  +- org.checkerframework:checker-qual:jar:3.12.0:compile
[INFO] |  |  \- com.google.j2objc:j2objc-annotations:jar:1.3:compile
[INFO] |  +- com.google.code.gson:gson:jar:2.8.9:compile
[INFO] |  +- com.google.protobuf:protobuf-java:jar:3.21.1:runtime
[INFO] |  +- com.google.protobuf:protobuf-java-util:jar:3.21.1:runtime
[INFO] |  \- io.grpc:grpc-netty-shaded:jar:1.38.0:runtime
[INFO] |     \- io.grpc:grpc-core:jar:1.38.0:runtime (version selected from constraint [1.38.0,1.38.0])
[INFO] |        +- com.google.android:annotations:jar:4.1.1.4:runtime
[INFO] |        \- io.perfmark:perfmark-api:jar:0.23.0:runtime
[INFO] +- com.pulumi:aws:jar:5.13.0:compile
[INFO] +- com.pulumi:kubernetes:jar:3.22.0-alpha.1661453084+94c160dd:compile
[INFO] \- com.pulumi:eks:jar:0.37.1:compile
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.653 s
[INFO] Finished at: 2022-08-29T10:44:07-04:00
[INFO] ------------------------------------------------------------------------
```

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### PR DESCRIPTION
**Pull request by pawelprazak**
# Description

Fixes #821 

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

**Comment by pawelprazak** [comment]
Working on fixing the inactive tests, but the fix and regression test looks good :)

**Comment by t0yv0** [comment]
Going to review a bit later when we get CI to green. Note that I looked into the example failure and it's unrelated, it was caused by the breaking changes propagating through azure-native provider, shortly fixed by https://github.com/pulumi/examples/pull/1275

File: pkg/codegen/java/build.gradle.template [hunk 1]
```diff
@@ -53,6 +53,17 @@ dependencies {
     {{- range $k, $v := .Dependencies }}
     implementation("{{ $k }}:{{ $v }}")
     {{- end }}
+
+    def junitVersion = "5.9.0"
+    testRuntimeOnly("org.junit.jupiter:junit-jupiter-engine:${junitVersion}")
+}
+
+test {
+    useJUnitPlatform()
```
Comment (pawelprazak) [inline pkg/codegen/java/build.gradle.template:62]: this re-enables the JUnit tests

File: pkg/codegen/testing/test/testdata/dashed-import-schema/java/src/main/java/com/pulumi/plant/outputs/Container.java [hunk 2]
```diff
@@ -71,7 +71,7 @@ public Builder material(@Nullable String material) {
         }
         @CustomType.Setter
         public Builder size(ContainerSize size) {
-            this.size = Objects.requireNonNull(size);
+            this.size = size;
```
Comment (pawelprazak) [inline pkg/codegen/testing/test/testdata/dashed-import-schema/java/src/main/java/com/pulumi/plant/outputs/Container.java:74]: the is the codegen result of the fix

File: pkg/codegen/testing/test/testdata/regress-821/java-extras/src/test/java/com/pulumi/aws/OutputTests.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+class OutputTests {
+
+    @AfterEach
+    void cleanup() {
+        PulumiTestInternal.cleanup();
+    }
+
+    @Test
+    void testPreviewOutput() {
```
Comment (pawelprazak) [inline pkg/codegen/testing/test/testdata/regress-821/java-extras/src/test/java/com/pulumi/aws/OutputTests.java:28]: the regression test

File: sdk/java/pulumi/src/test/java/com/pulumi/core/internal/UrnTest.java [hunk 4]
```diff
@@ -110,6 +110,19 @@ void testCreateUrnString(
         assertThat(urn).isEqualTo(expected);
     }
 
+    public static Stream<Arguments> testTypeParser() {
+        return Stream.of(
+                arguments("jumbo::ACustomResource", "jumbo:ACustomResource") // tolerate an extra ':'
```
Comment (pawelprazak) [inline sdk/java/pulumi/src/test/java/com/pulumi/core/internal/UrnTest.java:115]: not sure if this is undocumented URN feature, but without this codegen test JumboTest fails with the schema it has

**Comment by pawelprazak** [comment]
@t0yv0 ready for review :)

**Comment by t0yv0** [comment]
There's lots of good things in here but it's a bit much for a single PR. I've lifted #838 and #837 out into separate PRs.

On the fix itself I'm wondering if the fix can be accomplished in SDK only without changing the codegen. That is if the null checks are left in on builder methods but the deserializer code stops calling builder methods with nulls. The null checks seem like they have their place for normal usage, perhaps we should leave them? The other reason this is attractive is that an SDK-only fix is much easier for us to roll out.

**Comment by pawelprazak** [comment]
Thank you!

> On the fix itself I'm wondering if the fix can be accomplished in SDK only without changing the codegen. That is if the null checks are left in on builder methods but the deserializer code stops calling builder methods with nulls. The null checks seem like they have their place for normal usage, perhaps we should leave them? The other reason this is attractive is that an SDK-only fix is much easier for us to roll out.

Yes, it should be possible, but the the converter would have to be made aware of the preview mode, adding complexity.
The fix removes complexity. The null checks were not present in the constructor based impl and they add weight to the codegen-ed code. Given the above I've concluded that romeoval is the simplest an the best solution in those circumstances.

Having said that, there is nothing stopping us with going the other way and making the converter aware of the preview mode.

**Comment by t0yv0** [comment]
@pawelprazak I don't think the problem is exclusive to preview mode.. There is this regression test that was masked by JUnit tests not running that has a similar failure pattern but another root cause entirely. In case of AWS sometimes the provider fails to send back required output data - marked required by the schema but not actually present. This is in preview=False mode. Still we need to handle that null.

**Comment by pawelprazak** [comment]
> @pawelprazak I don't think the problem is exclusive to preview mode.. There is this regression test that was masked by JUnit tests not running that has a similar failure pattern but another root cause entirely. In case of AWS sometimes the provider fails to send back required output data - marked required by the schema but not actually present. This is in preview=False mode. Still we need to handle that missing data gracefully.

Right! Maybe it could be as simple as adding `return null;` here 🤔 :
Screenshot 2022-09-13 at 15 54 24

`com/pulumi/serialization/internal/Converter.java:423`

**Comment by pawelprazak** [comment]
> How about #840 ? It feels like the fix is falling at a somewhat strange place but it fixes both the broken example and the skipped test.

I've commented on the PR, TLDR: it should work fine.

File: sdk/java/pulumi/src/main/java/com/pulumi/serialization/internal/Converter.java [hunk 5]
```diff
@@ -421,24 +421,25 @@ private Object tryConvertObjectInner(
...
-                    throw new IllegalStateException(String.format(
-                            "Error invoking setter '%s' (on '%s'), setter parameters: '%s', argument type: '%s'",
-                            name, targetType.getTypeName(),
-                            Arrays.toString(setters.get(name).getParameterTypes()),
-                            argument == null ? "null" : argument.getClass()
-                    ), e);
-                } catch (IllegalAccessException | InvocationTargetException e) {
-                    var exMsg = e.getMessage() == null ? e.getClass().getSimpleName() : e.getMessage();
-                    throw new IllegalStateException(String.format("Unexpected exception: %s", exMsg), e);
+                } else {
```
Comment (pawelprazak) [inline sdk/java/pulumi/src/main/java/com/pulumi/serialization/internal/Converter.java:424]: @t0yv0  adding this else here, has a slightly different semantics then https://github.com/pulumi/pulumi-java/pull/840, but both should work in practice

the difference is, should we call the `@Nullable` setters, to make sure we don't have dirty mutable state issues, but on the other hand this else only helps with that, but do not guarantee lack of dirty state

**Comment by t0yv0** [comment]
@pawelprazak we ended up shipping #840 in the 0.6.0 release yesterday. Please let me know if we can close this PR or fi you think there's remaining value here I'm happy to take another look once it's rebased on main.

**Comment by pawelprazak** [comment]
thank you, I'll have a look at what minor improvements can be used after rebase

**Comment by t0yv0** [comment]
Closing for now as it trips our PR tracking tools. Please feel free to reopen if helpful when rebased.

**Pull request closed without merging**